### PR TITLE
cdc: rollback txn-source attached transactions correctly

### LINKED ISSUE #17211: rollback a txn_source specified transaction may blocks cdc
### What version of TiKV are you using?

The problem is introduced in #16645 .

See also https://github.com/pingcap/tiflow/issues/11341.

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Issue Number: Close #17211 

The problem is introduced in #16645 . This PR wants to allow advancing ResolvedTs if locks are filtered out by `txn_source`.

For example, if there is a transaction with start_ts `10` on a given region, and the transaction is performed with a `txn_source`, and a changefeed subscription is specified with `filter_loop = true`, ResolvedTs should not be blocked by the transaction.

However, we can't handle this situation: the transaction is rollbacked **without** a specified `txn_source`. This causes https://github.com/pingcap/tiflow/issues/11341.

So this PR fixes the problem by giving up the enhancement descriped above.

What's Changed:

```commit-message
cdc: rollback txn-source attached transactions correctly
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/delegate.rs [hunk 1]
```diff
@@ -951,12 +938,21 @@ impl Delegate {
                         }
                         BTreeMapEntry::Occupied(mut x) => {
                             *x.get_mut() += *lock_count_modify;
+                            assert!(*x.get() >= 0, "lock_count_modify should never be negative");
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:941]:
```suggestion
                            assert!(*x.get() >= 0, "lock_count_modify should never be negative, start_ts: {}", entry.start_ts);
```

File: components/cdc/src/delegate.rs [hunk 2]
```diff
@@ -1053,9 +1050,11 @@ impl Delegate {
     fn sink_delete(&mut self, mut delete: DeleteRequest, rows: &mut RowsBuilder) -> Result<()> {
         match delete.cf.as_str() {
             "lock" => {
-                if self.pop_lock(Key::from_encoded_slice(&delete.key))? != 0 {
+                if ApiV2::parse_key_mode(delete.get_key()) != KeyMode::Raw {
```
Comment (pingyu) [inline components/cdc/src/delegate.rs:1053]: We don't need this check. RawKV (API v2, and only API v2 can use CDC) has no lock and will write to default cf only.

**Review by pingyu — APPROVED** [review]
LGTM~

**Pull request merged**